File: feed/fetcher.go
package feed

import (
	"context"
	"fmt"
	"net/http"
	"sort"
	"sync"
	"time"

	"github.com/mmcdole/gofeed"
)

type Post struct {
	FeedName    string    `json:"feed_name"`
	Title       string    `json:"title"`
	URL         string    `json:"url"`
	PublishedAt time.Time `json:"published_at"`
}

type FeedError struct {
	FeedName string
	URL      string
	Err      error
}

func (e FeedError) Error() string {
	return fmt.Sprintf("feed %q (%s): %v", e.FeedName, e.URL, e.Err)
}

func FetchFeed(ctx context.Context, feedURL string, maxPosts int) ([]Post, error) {
	parser := gofeed.NewParser()
	parser.Client = &http.Client{
		Transport: http.DefaultTransport,
	}

	f, err := parser.ParseURLWithContext(feedURL, ctx)
	if err != nil {
		return nil, err
	}

	var posts []Post
	for i, item := range f.Items {
		if i >= maxPosts {
			break
		}
		published := time.Time{}
		if item.PublishedParsed != nil {
			published = item.PublishedParsed.UTC()
		} else if item.UpdatedParsed != nil {
			published = item.UpdatedParsed.UTC()
		}

		posts = append(posts, Post{
			FeedName:    f.Title,
			Title:       item.Title,
			URL:         item.Link,
			PublishedAt: published,
		})
	}
	return posts, nil
}

func ValidateFeed(ctx context.Context, url string) error {
	_, err := FetchFeed(ctx, url, 1)
	return err
}

func FetchAllFeeds(ctx context.Context, feeds []Feed, maxPostsPerFeed int, timeout time.Duration) ([]Post, []FeedError) {
	var (
		mu       sync.Mutex
		allPosts []Post
		errs     []FeedError
		wg       sync.WaitGroup
	)

	for _, f := range feeds {
		wg.Add(1)
		go func(fd Feed) {
			defer wg.Done()
			fetchCtx, cancel := context.WithTimeout(ctx, timeout)
			defer cancel()

			posts, err := FetchFeed(fetchCtx, fd.URL, maxPostsPerFeed)
			mu.Lock()
			defer mu.Unlock()
			if err != nil {
				errs = append(errs, FeedError{FeedName: fd.Name, URL: fd.URL, Err: err})
				return
			}
			// Override feed name with the user-configured name
			for i := range posts {
				posts[i].FeedName = fd.Name
			}
			allPosts = append(allPosts, posts...)
		}(f)
	}

	wg.Wait()

	sort.Slice(allPosts, func(i, j int) bool {
		return allPosts[i].PublishedAt.After(allPosts[j].PublishedAt)
	})

	return allPosts, errs
}

File: src/feed/fetcher.py
"""RSS/Atom feed fetcher using httpx and feedparser."""

import asyncio
from datetime import datetime, timezone

import feedparser
import httpx

from feed.models import Feed, FeedError, Post

EPOCH = datetime(1970, 1, 1, tzinfo=timezone.utc)


class HttpFeedFetcher:
    def __init__(self, transport: httpx.AsyncBaseTransport | None = None):
        self._transport = transport

    def _parse_timestamp(self, entry) -> datetime:
        if entry.get("published_parsed"):
            return datetime(*entry.published_parsed[:6], tzinfo=timezone.utc)
        if entry.get("updated_parsed"):
            return datetime(*entry.updated_parsed[:6], tzinfo=timezone.utc)
        return EPOCH

    async def fetch_feed(
        self, feed_name: str, feed_url: str, max_posts: int
    ) -> list[Post]:
        async with httpx.AsyncClient(transport=self._transport) as client:
            response = await client.get(feed_url)
            response.raise_for_status()
            content = response.text

        parsed = await asyncio.to_thread(feedparser.parse, content)
        if not parsed.version:
            raise ValueError(f"URL '{feed_url}' is not a valid feed")

        posts = []
        for entry in parsed.entries[:max_posts]:
            posts.append(
                Post(
                    feed_name=feed_name,
                    title=entry.get("title", ""),
                    url=entry.get("link", ""),
                    published_at=self._parse_timestamp(entry),
                )
            )
        return posts

    async def validate_feed(self, name: str, url: str) -> None:
        await self.fetch_feed(name, url, max_posts=1)

    async def fetch_all_feeds(
        self, feeds: list[Feed], max_posts_per_feed: int, timeout: float
    ) -> tuple[list[Post], list[FeedError]]:
        async def fetch_one(feed: Feed) -> tuple[list[Post], FeedError | None]:
            try:
                posts = await asyncio.wait_for(
                    self.fetch_feed(feed.name, feed.url, max_posts_per_feed),
                    timeout=timeout,
                )
                return posts, None
            except Exception as exc:
                return [], FeedError(feed_name=feed.name, url=feed.url, error=str(exc))

        results = await asyncio.gather(*(fetch_one(f) for f in feeds))

        all_posts: list[Post] = []
        all_errors: list[FeedError] = []
        for posts, error in results:
            all_posts.extend(posts)
            if error is not None:
                all_errors.append(error)

        all_posts.sort(key=lambda p: p.published_at, reverse=True)
        return all_posts, all_errors
